libraries/AP_UAVCAN/AP_UAVCAN.h

```diff
@@ -32,253 +37,178 @@
 #define UAVCAN_SRV_NUMBER 18
 #endif
 
-#define AP_UAVCAN_MAX_LISTENERS 4
-#define AP_UAVCAN_MAX_GPS_NODES 4
-#define AP_UAVCAN_MAX_MAG_NODES 4
-#define AP_UAVCAN_MAX_BARO_NODES 4
-#define AP_UAVCAN_MAX_BI_NUMBER 4
-
 #define AP_UAVCAN_SW_VERS_MAJOR 1
 #define AP_UAVCAN_SW_VERS_MINOR 0
 
 #define AP_UAVCAN_HW_VERS_MAJOR 1
 #define AP_UAVCAN_HW_VERS_MINOR 0
 
 #define AP_UAVCAN_MAX_LED_DEVICES 4
-#define AP_UAVCAN_LED_DELAY_MILLISECONDS 50
 
-class AP_UAVCAN {
+/*
+    Frontend Backend-Registry Binder: Whenever a message of said DataType_ from new node is received,
+    The Callback will invoke frontend to register the node as separate backend.
+
+    Note: Generic Subscriptions should not be done using this mechanism, i.e. messages for which
+    you don't want Frontend to init a new Backend when they come from different Nodes.
+    Use uavcan::MethodBinder or util/Functor template instead and initialise the subscriber in the Backend Constructor.
+*/
+#define UC_FRONTEND_REGISTRY_BINDER(ClassName_, DataType_, Backend_, Frontend_)\
```

Comment:
do these two macros need to be like this? Can they be a bit simpler and call into some ordinary C++ code instead?

Reply:
There needs to be a class with overload function operator with given templated argument, to be passed down as functor  https://github.com/ArduPilot/ardupilot/pull/8972/commits/977caee4ce286f98ec260130408ab46d4c537eda#diff-ceee283277029c29c750143f987a85c7R66 . I wanted these to be template but forward declaring templates with uavcan message typename would not be possible as they are typedefs. This is the best solution I could come up with.

Reply:
>  forward declaring templates with uavcan message typename would not be possible as they are typedefs.

This is definitely possible, we talked about it in the dev call yesterday... I can post it here again if you want.